pulp_container/tests/functional/api/test_repositories_list.py

```diff
@@ -30,8 +32,24 @@
 )
 
 
-class RepositoriesListTestCase(unittest.TestCase):
-    """Test case for listing all repositories within a registry."""
+class NoRebuildAuthSession(requests.Session):
+    """A dummy session wrapper which do not change the authentication headers during redirects.
+
+    According to the docs https://buildmedia.readthedocs.org/media/pdf/requests/master/requests.pdf,
+    authorization headers will be removed if you get redirected off-host. In order to override such
+    behaviour, it is necessary to redefine the method which is responsible for changing the
+    authentication header.
+
+    This 'hack' was required for running the tests in the containerized environment, where redirects
+    happen. Credits to https://stackoverflow.com/a/60451840/3907906.
+    """
+
+    def rebuild_auth(self, prepared_request, response):
+        """Do not change the authentication header."""
```

Comment:
I think i need more info here. Who is redirecting where?
This "container setup" is a possible real world deployment, so this must not be needed.

Reply:
So, it behaved like this: When I removed the `.netrc` file, the tests were passing. When I disabled redirects by passing `allow_redirects=False` to the get method, it did not reach the token server. When I used a custom session object (i.e., `session.get`) instead of `requests` (`requests.get`), the tests were passing even when the `.netrc` file was defined.

I thought that this behaviour could be caused either because of "Authorization headers will be removed if you get redirected off-host" or "Authorization headers set with headers= will be overridden if credentials are specified in .netrc, which in turn will be overridden by the auth= parameter. Requests will search for the netrc file at ~/.netrc, ~/_netrc, or at the path specified by the NETRC environment variable." (see Section 2.2.7 https://buildmedia.readthedocs.org/media/pdf/requests/master/requests.pdf). I decided to consider just the first statement.

Reply:
My question still: Why is there a redirect at all? What is different in the other tests where we don't need such hackery?
Does the first question mean we are deploying pulp wrongly in the ci?

Reply:
So, `pulp:80/token` is redirected to `pulp/token/` because the Authorization header contains the following value:

`Bearer realm="http://pulp:80/token",service="pulp",scope="registry:catalog:*"`

Could be this discrepancy caused by the configuration declared in `pulp_container/app/webserver_snippets/*.conf`? Would changing `/token http://${pulp-api}/token` to `/token/ http://${pulp-api}/token/` resolve the problem?

Reply:
Actually, changing the `TOKEN_SERVER` setting will resolve the problem.

Reply:
Wait, this is redirection by the reverse proxy. The client shouldn't even see this. This is no explanation that we need to mess with the requests library. And still we are calling to the token api in other tests too.
Or are you saying that the client visible redirect is from "http://pulp:80/token" to "http://pulp:80/token/"?

Reply:
Yes, the client is redirected from `/token` to `/token/`. We are calling the token API without providing authentication headers. That is why it worked in other cases.

Reply:
Can you add the / in the authentication header?